drivers/power_domain/power_domain_intel_adsp.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(power_domain_intel_adsp, LOG_LEVEL_INF);
+
+struct pg_registers {
+	uint32_t wr_address;
+	uint32_t rd_address;
+	uint32_t SPA_bit;
+	uint32_t CPA_bit;
+};
+
+static int pd_intel_adsp_set_power_enable(struct pg_registers *reg, bool power_enable)
+{
+	uint32_t SPA_bit_mask = BIT(reg->SPA_bit);
+
+	if (power_enable) {
+		sys_write32(sys_read32(reg->wr_address) | SPA_bit_mask, reg->wr_address);
+	} else {
+		sys_write32(sys_read32(reg->wr_address) & ~(SPA_bit_mask), reg->wr_address);
+	}
+
+	return 0;
+}
+
+static int pd_intel_adsp_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	struct pg_registers *reg_data = (struct pg_registers *)dev->data;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_ON, NULL);
+		pd_intel_adsp_set_power_enable(reg_data, true);
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_OFF, NULL);
+		pd_intel_adsp_set_power_enable(reg_data, false);
+		break;
+	case PM_DEVICE_ACTION_TURN_ON:
+		break;
+	case PM_DEVICE_ACTION_TURN_OFF:
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+static int pd_intel_adsp_init(const struct device *dev)
+{
+	pm_device_init_suspended(dev);
+	pm_device_runtime_enable(dev);
+	return 0;
+}
+
+#define DT_DRV_COMPAT intel_adsp_power_domain
+
+#define POWER_DOMAIN_DEVICE(id)                                                                    \
+	static struct pg_registers pd_pg_reg##id = {                                               \
+		.wr_address = DT_INST_PROP(id, write_address),                                     \
+		.rd_address = DT_INST_PROP(id, read_address),                                      \
+		.SPA_bit = DT_INST_PROP(id, bit_position),                                         \
+		.CPA_bit = DT_INST_PROP(id, bit_position),                                         \
```

Comment:
This seems clumsy.  The PWRCTL register is part of the shim block, which already has a devicetree definition. So you can't declare an overlapping register block, which is presumably why you're using DTS properties to store the address instead?

@mbolivar can surely provide better guidance than I can, but this seems like a no-no in DTSese.  I think the proper way to do this would be to declare a symbolic reference to the other device to which you could follow and then get its registers.

But really that's sort of silly, as this is a VERY simple device, and there's an existing CAVS_SHIM abstraction in the soc headers (and, again, a similar one for MTL) that you can use just like a C struct to refer to the PWRCTL register.  If you want the address of PWRCTL to pass to sys_read/write(), then it's just `&CAVS_SHIM.pwrctl`.

Please unify.  I don't care if you keep the existing API (I like it a lot, but I don't maintain it anymore) or go with a more traditional one, but we don't want two DTS blocks owning the same piece of hardware.